### DISCUSSION TITLE
How to create a USI (Unified System Images which)?

### DISCUSSION THREAD

**Discussion by septatrix** [systemd/mkosi#2349] [Q&A] [answered]
The recent blog post "[A re-introduction to mkosi](https://0pointer.net/blog/a-re-introduction-to-mkosi-a-tool-for-generating-os-images.html)" talked about USIs (Unified System Images which are full OS images packed in a UKI) though they are not a first-class output format (i.e. one cannot say `mkosi --format usi`. The systemd or UAPI projects also do not seem to document what exactly constitutes a USI.

Is it "just" a normal system (i.e. `os-release` instead of `initrd-release`) bundled with the kernel similar to a traditional initrd? If this is the case how is this handled by systemd? Is there still an initrd stage for early-boot units anymore etc...?

**Comment by Winterhuman** [comment, marked as answer]
I have a blog post about building USIs with `mkosi`: https://overhead.neocities.org/blog/build-usi-mkosi

But TLDR: Use `Format=uki` plus `MakeInitrd=no` to create a USI.

**Reply by septatrix** [reply]
Thanks, that was an interesting read. Did you run into some problems due to services not being run which only run during the initrd phase? I was thinking about stuff like `systemd-pcrphase-initrd.service` etc.

PS: Maybe consider adding syntax highlighting to your blog for the systemd files (INI syntax should be sufficient)

**Reply by Winterhuman** [reply]
`systemd-pcrphase-initrd.service` is inactive for me, but generally speaking, not creating `/etc/initrd-release` is all systemd needs to understand that the UKI isn't an initramfs

**Reply by edorgeville** [reply]
Very interesting read, thanks @Winterhuman

**Reply by digitalsignalperson** [reply]
Hey @Winterhuman, I'm trying your exact `mkosi.conf` but it is not bootable. i.e. `systemd-vmspawn --register=no --linux=example-usi.efi` doesn't work. I'll get these errors:

```
BdsDxe: failed to load Boot0002 "UEFI Non-Block Boot Device" from VenMedia(1428F772-B64A-441E-B8C3-9EBDD7F893C7): Not Found
BdsDxe: failed to load Boot0003 "UEFI Non-Block Boot Device 2" from PciRoot(0x0)/Pci(0x5,0x0): Not Found
BdsDxe: No bootable option or device was found.
BdsDxe: Press any key to enter the Boot Manager Menu.
```

I can confirm being able to boot other hand made arch linux USIs with `systemd-vmspawn` in this way. My custom USIs were made with ukify, but with outputs from mkinitcpio, including a custom initcpio hook to setup some things.

Any ideas what is missing?

**Reply by Winterhuman** [reply]
@digitalsignalperson I'm not too surprised, since it's a pretty old `mkosi` config, though I couldn't tell you off the top of my head what needs changing; I only ever booted that config with the old Qemu VM `mkosi` used to use
